fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    fetch('/api/todos')
+      .then(res => res.json())
+      .then(todos => this.setState({
+        todos
+      }));
+  }
+
+  addTodo(newTodo) {
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(todo => {
+        this.state.todos.push(todo);
+        return this.setState({
+          todos: this.state.todos
+        });
+      });
+
+  }
+
+  toggleCompleted(todoId) {
+    const todoList = this.state.todos;
+    const result = todoList.map(todo => {
+      if (todo.todoId === todoId) {
+        return { isCompleted: todo.isCompleted };
+      }
+    });
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(result[todoId - 1])
+    })
+      .then(res => res.json())
+      .then(data => {
+        const current = this.state.todos[todoId - 1];
+        current.isCompleted = !data.isCompleted;
+        return this.setState({
+          todos: this.state.todos
+        });
```

Comment:
Instead of reaching into your state, you should
1. make a copy of the original array with `slice()`.
2. Assign the `data` you got from the server to the appropriate index.
3. setState

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice